### ISSUE TITLE
Adding groups to pointers?

### ISSUE THREAD

**Issue by interior-night** [frauzufall/ofxGuiExtended#39]
Not sure if this is an issue but I'm using ofxAppUtils with ofxGuiExtended and have the gui set up in my ofApp.cpp. I can add panels to that gui from a _scene_ like so:
`scene08Panel = mainApp->gui.addPanel("Scene 08");`
But my app crashes on launch if I try to add a group defined in a scene to a panel defined in the ofApp: `scene08Group = mainApp->scenePanel->addGroup("Scene 08");` Am I doing this wrong?

**Comment by frauzufall** [comment]
This should work. Are you sure the second line gets called after the first line? Because it will crash if `scenePanel` is still a null pointer.

**Comment by interior-night** [comment]
I'm only trying to use that second line in my scene. The gui itself is defined in my ofApp... so shouldn't be a null pointer I presume?

**Comment by interior-night** [comment]
This is some of what xcode spits out, if it helps:

```
->  0x1001a6550 <+832>:  movq   0x8(%rcx), %r14
    0x1001a6554 <+836>:  movq   %rcx, %r15
    0x1001a6557 <+839>:  movq   %r15, -0x208(%rbp)
    0x1001a655e <+846>:  movq   -0x208(%rbp), %r15
    0x1001a6565 <+853>:  addq   $0x10, %r15
    0x1001a6569 <+857>:  movq   %r15, -0x200(%rbp)
    0x1001a6570 <+864>:  movq   -0x200(%rbp), %r15
    0x1001a6577 <+871>:  movq   %r15, -0x1f8(%rbp)
    0x1001a657e <+878>:  movq   -0x1f8(%rbp), %r15
    0x1001a6585 <+885>:  cmpq   (%r15), %r14
    0x1001a6588 <+888>:  movq   %rcx, -0x5d8(%rbp)
    0x1001a658f <+895>:  jae    0x1001a683f               ; <+1583> [inlined] std::__1::vector<std::__1::unique_ptr<DOM::Element, std::__1::default_delete<DOM::Element> >, std::__1::allocator<std::__1::unique_ptr<DOM::Element, std::__1::default_delete<DOM::Element> > > >::push_back(std::__1::unique_ptr<DOM::Element, std::__1::default_delete<DOM::Element> >&&) + 751 at Element.h:688
```

**Comment by frauzufall** [comment]
Nah no idea what's going wrong. I did not work with ofxAppUtils before. 
Do you create the scenePanel in ofApp before you add the sceneGroup in your scene class to the panel? 
If that's not the issue, could you give me access to the code or a minimal example so I can debug this?

**Comment by interior-night** [comment]
Cool. I've sent you a link to a minimal example through the oF forums. It should work as is (at least on a mac) but in scene08.h, for example, if you change

```
        scene08Panel = mainApp->gui.addPanel("Scene 08", ofJson({{"flex-direction", "row"}, {"flex-wrap", "wrap"}}));
        scene08Panel->loadTheme("assets/themes/theme_bleurgh.json");

        scene08Panel->setPosition(ofGetWidth()-330, 0);
        scene08Panel->setShowHeader(false);

        scene08Group = scene08Panel->addGroup("Scene 08");

        scene08Group->add(sphereRadius);
        scene08Group->add(sphereResolution);
        scene08Group->add(isFilled);
        scene08Group->add(isWireframe);
        scene08Group->add(hasNormalsDrawn);
        scene08Group->add(hasAxesDrawn);
        scene08Group->add(hasLightsDrawn);
        scene08Group->add(hasFacesSplit);

        // scene's panel starts off hidden
        scene08Panel->getVisible().set(false);
```

to

```
        scene08Group = mainApp->menuScenePanel->addGroup("Scene 08");

        scene08Group->add(sphereRadius);
        scene08Group->add(sphereResolution);
        scene08Group->add(isFilled);
        scene08Group->add(isWireframe);
        scene08Group->add(hasNormalsDrawn);
        scene08Group->add(hasAxesDrawn);
        scene08Group->add(hasLightsDrawn);
        scene08Group->add(hasFacesSplit);

        // scene's panel starts off hidden
        scene08Group->getVisible().set(false);
```

as well as changing all references of scene08Panel to scene08Group then it no longer works.

By the way, left/right arrow keys navigate from one scene to the next.

**Comment by frauzufall** [comment]
ofxAppUtils seems to depend on ofxGui. Do you have a version that uses ofxGuiExtended or how are you compiling this?

**Comment by interior-night** [comment]
Yeah, ofxAppUtils does depend on ofxGui for its 'transformer' but I've disabled that for now and will try to make it ofxguiextended compatible soon. The transformer is a bit like a lesser version of your ofx2dMapping, by the way. I'd rather use yours instead but need the scenemanager functionality that ofxAppUtils provides (basically, scenes with fade/crossfade transitions).

**Comment by frauzufall** [comment]
Alright.

You call `setupGui()` in line 109 in your ofApp. That's after you add your scenes from line 58 on. That means the main panel is created after you want to add things to it in the scene. This won't work. Just move setupGui() to a place before you create the scenes.
That's also the reason why your horizontal panels are not working: You are using the flexbox layout and by calling gui.setupFlexBoxLayout() which needs to be done before you do anything else with the gui. Look at the error messages, there is also one telling you that.

**Comment by frauzufall** [comment]
Also if you want to adapt the transormer to ofxGuiExtended you can start with this branch https://github.com/frauzufall/ofxAppUtils/tree/ofxguiextended I just rewrote it to using ofxGuiExtended but I have no idea how to test it. But it could be a useful starting point.

**Comment by interior-night** [comment]
Whoops. Working now (though scene08Group is not inheriting the styling of menuScenePanel... but haven't tried to troubleshoot that yet!) Thanks for having a look for me! You've been awesome.

I'll have a look at your ofxapputils branch right now (or maybe right after i watch the latest episode of game of thrones...) Will report back.

**Comment by frauzufall** [comment]
:) Since all of this is pretty new it is harder to tell on which side of the code a problem occurs.. But also thanks to your tests this will get more reliable with time.

**Comment by interior-night** [comment]
Yep. Once you’ve got all my issues resolved you’ll be able to say your gui is officially fool-proof! But don’t underestimate my foolishness!

**Comment by interior-night** [comment]
Your branch of ofxAppUtils crashes on my mac with the following, for what it's worth...

`0x100348688 <+872>: jmp    0x10034870b               ; nlohmann::basic_json<std::__1::map, std::__1::vector, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool, long long, unsigned long long, double, std::__1::allocator>& nlohmann::basic_json<std::__1::map, std::__1::vector, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, bool, long long, unsigned long long, double, std::__1::allocator>::operator[]<char const>(char const*) + 1003`

**Comment by interior-night** [comment]
![screen shot 2016-05-23 at 22 17 53](https://cloud.githubusercontent.com/assets/652507/15470487/5d26cb1e-2134-11e6-9884-00f469f187f7.png)

**Comment by frauzufall** [comment]
yes sorry try again (-> pull). is there an example where a panel is used? the standard example of ofxAppUtils shows nothing. how is this supposed to work?

**Comment by frauzufall** [comment]
btw, in my version the panel has its own ofxGui object it is attached to. If this should be connected to an existing gui (and get its theme etc.) this as to be done differently(like giving the panel a parent panel it can attach itself to with the setup function)

**Comment by interior-night** [comment]
Back from Game of Thrones! Awesome - your last fix got it working. To test it out, btw, `#define HAVE_OFX_GUI` needs to be uncommented at the top of ofApp.h and then i think it's a question of hitting the 'd' key when viewing a scene to bring up the transformer.

**Comment by interior-night** [comment]
So, in my case, if i want to attach the transform panel to a pre-exisiting ofApp panel, can i change
`panel = gui.addPanel(name, ofJson({{"filename", ofToDataPath(xmlFile)}}));`
to
   `panel = mainApp->menuTransformGroup;` in ofxTransformPanel.cpp, and

```
ofxGui gui;
ofxGuiPanel *panel; //< gui panel
```

to

`ofxGuiGroup *panel;` in the transform header? And if so, how do I create that "mainApp" pointer to the ofApp panel? From a scene, I can create such a pointer with `ofApp *mainApp = (ofApp *)ofxGetAppPtr();` but that doens't work in the transformpanel's header.

**Comment by frauzufall** [comment]
Can you write something like `panel = (ofApp *)ofxGetAppPtr()->menuTransformGroup->addPanel(..)`?

**Comment by interior-night** [comment]
`(ofApp *)ofxGetAppPtr()` seems to refer to ofxApp rather than ofApp. And if I try `(ofApp *)ofGetAppPtr()` then it refers to ofBaseApp. So still can't reach menuTransformGroup...

**Comment by interior-night** [comment]
`panel = (ofApp *)ofxGetAppPtr()->menuTransformGroup->addPanel("test");` gives me the following error: **use of undeclared identifier 'ofApp'** and I can't figure out how to declare it in the header.

**Comment by frauzufall** [comment]
did you include "ofApp.h" in the panel .cpp file? Why would you want to declare it in the header? Also, try to wrap the app cast in braces: `((ofApp *)ofxGetAppPtr())->menuTransformGroup->addPanel("test");`
Sorry I can only give you some hints right now, if you still have problems I can test it later.

**Comment by interior-night** [comment]
Yeah. I had tried that to declare it in the .cpp and that's when i get the **no member named 'menuTransformGroup' in ofxApp** error instead. But wrapping the cast in braces works! The app builds and then crashes with the error message from way at the top of this thread. So i'm guessing I need to check the order of things, that the transformer panel is not being created before the rest of the gui. I'll investigate.  I think I started on openframeworks and c++ less than a month ago and I've learned a lot thanks to your help. It's much appreciated!
